pkg/controller/deployment_inplace.go

```diff
@@ -0,0 +1,584 @@
+		}
+
+		node, err := dc.nodeLister.Get(nodeName)
+		if err != nil {
+			return fmt.Errorf("failed to get node %s: %w", nodeName, err)
+		}
+
+		cond := nodeops.GetCondition(node, v1alpha1.NodeInPlaceUpdate)
+		if isUpdateNotSuccessful(cond, node.Labels) {
+			continue
+		}
+
+		// remove labels related to the inplace update.
+		delete(node.Labels, v1alpha1.LabelKeyNodeCandidateForUpdate)
+		delete(node.Labels, v1alpha1.LabelKeyNodeSelectedForUpdate)
+		delete(node.Labels, v1alpha1.LabelKeyNodeUpdateResult)
+		// remove annotations related to the inplace update.
+		delete(node.Annotations, v1alpha1.AnnotationKeyMachineUpdateFailedReason)
+
+		// uncordon the node since the inplace update is successful.
+		node.Spec.Unschedulable = false
+
+		_, err = dc.targetCoreClient.CoreV1().Nodes().Update(ctx, node, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to remove inplace labels/annotations and uncordon node %s: %w", node.Name, err)
+		}
+	}
+
+	for _, machineSet := range oldMachineSets {
+		// scale down the old machine set to the number of machines which is having the labelselector of the machine set.
+		oldMachines, err := dc.machineLister.List(labels.SelectorFromSet(machineSet.Spec.Selector.MatchLabels))
+		if err != nil {
+			return fmt.Errorf("failed to list machines for machine set %s: %w", machineSet.Name, err)
+		}
+
+		if len(oldMachines) < int(machineSet.Spec.Replicas) {
+			_, _, err := dc.scaleMachineSetAndRecordEvent(ctx, machineSet, int32(len(oldMachines)), deployment) // #nosec G115 (CWE-190) -- value already validated
+			if err != nil {
+				return fmt.Errorf("failed to scale down machine set %s: %w", machineSet.Name, err)
+			}
+			klog.V(3).Infof("scaled down machine set %s to %d", machineSet.Name, len(oldMachines))
+		}
+	}
+
+	return nil
+}
+
+func (dc *controller) reconcileNewMachineSetInPlace(ctx context.Context, oldMachineSets []*v1alpha1.MachineSet, newMachineSet *v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment) (bool, error) {
+	if newMachineSet.Spec.Replicas == deployment.Spec.Replicas {
+		// Scaling not required.
+		return false, nil
+	}
+
+	if newMachineSet.Spec.Replicas > deployment.Spec.Replicas {
+		// Scale down.
+		scaled, _, err := dc.scaleMachineSetAndRecordEvent(ctx, newMachineSet, deployment.Spec.Replicas, deployment)
+		return scaled, err
+	}
+
+	klog.V(3).Infof("reconcile new machine set %s", newMachineSet.Name)
+
+	oldMachinesCount := GetReplicaCountForMachineSets(oldMachineSets)
+	if oldMachinesCount == 0 {
+		scaled, _, err := dc.scaleMachineSetAndRecordEvent(ctx, newMachineSet, deployment.Spec.Replicas, deployment)
+		if err != nil {
+			return false, fmt.Errorf("failed to scale up machine set %s: %w", newMachineSet.Name, err)
+		}
+		klog.V(3).Infof("scaled up machine set %s to %d", newMachineSet.Name, deployment.Spec.Replicas)
+		return scaled, err
+	}
+
+	addedNewReplicasCount := int32(0)
```

Comment:
you can just defined this as `	var addedNewReplicasCount int32` instead of using casting to tell golang compiler the type of the variable. Default value for int32 is anyways 0.